src/codegen/llvm/replace_with_lib_functions.cpp

```diff
@@ -0,0 +1,214 @@
+            tli.addVectorizableFunctions(x86_functions);
+        }
+
+    } else {
+        // A map to query vector library by its string value.
+        using VecLib = TargetLibraryInfoImpl::VectorLibrary;
+        static const std::map<std::string, VecLib> llvm_supported_vector_libraries = {
+            {"Accelerate", VecLib::Accelerate},
+            {"libmvec", VecLib::LIBMVEC_X86},
+            {"libsystem_m", VecLib ::DarwinLibSystemM},
+            {"MASSV", VecLib::MASSV},
+            {"none", VecLib::NoLibrary},
+            {"SVML", VecLib::SVML}};
+
+        const auto& library = llvm_supported_vector_libraries.find(platform->get_math_library());
+        if (library == llvm_supported_vector_libraries.end())
+            throw std::runtime_error("Error: unknown vector library - " + platform->get_math_library() + "\n");
+
+        // Add vectorizable functions to the target library info.
+        switch (library->second) {
+        case VecLib::LIBMVEC_X86:
+            if (!triple.isX86() || !triple.isArch64Bit())
+                break;
+        default:
+            tli.addVectorizableFunctionsFromVecLib(library->second);
+            break;
+        }
+    }
+}
+
+void ReplaceWithLibdevice::getAnalysisUsage(AnalysisUsage& au) const {
+  au.setPreservesCFG();
+  au.addPreserved<ScalarEvolutionWrapperPass>();
+  au.addPreserved<AAResultsWrapperPass>();
+  au.addPreserved<LoopAccessLegacyAnalysis>();
+  au.addPreserved<DemandedBitsWrapperPass>();
+  au.addPreserved<OptimizationRemarkEmitterWrapperPass>();
+  au.addPreserved<GlobalsAAWrapperPass>();
+}
+
+bool ReplaceWithLibdevice::runOnFunction(Function& function) {
+    bool modified = false;
+
+    // Try to replace math intrinsics.
+    std::vector<CallInst*> replaced_calls;
+    for (auto& instruction: instructions(function)) {
+        if (auto* call_inst = dyn_cast<CallInst>(&instruction)) {
+            if (replace_call(*call_inst)) {
+                replaced_calls.push_back(call_inst);
+                modified = true;
+            }
+        }
+    }
+
+    // Remove calls to replaced intrinsics.
+    for (auto* call_inst: replaced_calls) {
+        call_inst->eraseFromParent();
+    }
+
+    return modified;
+}
+
+bool ReplaceWithLibdevice::replace_call(CallInst& call_inst) {
+    Module* m = call_inst.getModule();
+    Function* function = call_inst.getCalledFunction();
+
+    // Replace math intrinsics only!
+    auto id = function->getIntrinsicID();
+    bool is_nvvm_intrinsic = id == Intrinsic::nvvm_read_ptx_sreg_ntid_x ||
+            id == Intrinsic::nvvm_read_ptx_sreg_nctaid_x ||
+            id == Intrinsic::nvvm_read_ptx_sreg_ctaid_x ||
+            id == Intrinsic::nvvm_read_ptx_sreg_tid_x;
+    if (id == Intrinsic::not_intrinsic || is_nvvm_intrinsic)
+        return false;
+
+    // Map of supported replacements. For now it is only exp.
+    static const std::map<std::string, std::string> libdevice_name = {
+            {"llvm.exp.f32", "__nv_expf"},
```

Comment:
I think it would be good to add also `pow` and maybe look for any other math function commonly used in the mod files and also add it to the `x86` and `aarch64` maps. I can look at this also the next days